gobblin-temporal/src/main/java/org/apache/gobblin/temporal/joblauncher/GobblinTemporalJobLauncher.java

```diff
@@ -113,7 +125,61 @@ protected void handleLaunchFinalization() {
 
   @Override
   protected void executeCancellation() {
-    log.info("Cancel temporal workflow");
+    if (this.workflowId == null) {
+      log.info("Cancellation of temporal workflow attempted without submitting it");
+      return;
+    }
+
+    log.info("Cancelling temporal workflow {}", this.workflowId);
+    try {
+      WorkflowStub workflowStub = this.client.newUntypedWorkflowStub(this.workflowId);
+
+      // Describe the workflow execution to get its status
+      DescribeWorkflowExecutionRequest request = DescribeWorkflowExecutionRequest.newBuilder()
+          .setNamespace(this.namespace)
+          .setExecution(workflowStub.getExecution())
+          .build();
+      DescribeWorkflowExecutionResponse response = workflowServiceStubs.blockingStub().describeWorkflowExecution(request);
+
+      WorkflowExecutionStatus status;
+      try {
+        status = response.getWorkflowExecutionInfo().getStatus();
+      }
+      catch (Exception e) {
+        log.warn("Exception occurred while getting status of the workflow " + this.workflowId
+            + ". We would still attempt the cancellation", e);
+        workflowStub.cancel();
+        log.info("Temporal workflow {} cancelled successfully", this.workflowId);
+        return;
+      }
+
+      // Check if the workflow is not finished
+      if (status != WorkflowExecutionStatus.WORKFLOW_EXECUTION_STATUS_COMPLETED &&
+          status != WorkflowExecutionStatus.WORKFLOW_EXECUTION_STATUS_FAILED &&
+          status != WorkflowExecutionStatus.WORKFLOW_EXECUTION_STATUS_CANCELED &&
+          status != WorkflowExecutionStatus.WORKFLOW_EXECUTION_STATUS_TERMINATED) {
+        workflowStub.cancel();
+        try {
+          // Check workflow status, if it is cancelled, will throw WorkflowFailedException else TimeoutException
+          workflowStub.getResult(TERMINATION_TIMEOUT_SECONDS, TimeUnit.SECONDS, String.class, String.class);
+        }
+        catch (TimeoutException te) {
+          // Workflow is still running, terminate it.
+          log.info("Workflow is still running, will attempt termination", te);
+          workflowStub.terminate("Job cancel invoked");
```

Comment:
Should we have a wait in between the cancel and fetching the status after cancellation and make that a parameter? Seems quite aggressive, unless we know that workflowStub.cancel will only complete if the cancel is fully done gracefully

Reply:
The getResult api returns only if the workflow is completed. If the workflow is running it will wait for 3 seconds and raise a TimeoutException. It is a blocking call till the time workflow is in running state, so we should be good here.

Reply:
I think it can still be a bit risky here, only waiting 3 seconds for the workflow to cancel doesn't make sense once we implement heartbeats on the activities. It seems like to handle this cleanly we will need to explore adding heartbeats and acks to each of our activities, please correct me if I'm missing something here. I'd rather we just request cancellation, and not forcefully terminate. If the job cannot cancel on its own at this stage, we would let the service operator force terminate the flow. At least given the current state of the system I think this would lead to less unexpected terminations and operator pain.

Reply:
When the AZ driver job is cancelled, it triggers a cancel on Yarn application master. The application master has about 5 seconds to do the cleanup, post which the Application master and the workers (containers) launched by it are killed. Performing cancellation requires active workers, so the workflow would remain in Running state forever (similar to current state).

Can you please elaborate on how termination can impact the flow for which a cancel request is being sent?